File: rustprogramming/hash/1_two_sum.rs
use std::collections::HashMap;

struct Solution;

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut map: HashMap<i32, usize> = HashMap::new();
        for i in 0..nums.len() {
            if map.contains_key(&nums[i]) {
                let last_index = map[&nums[i]];
                return vec![last_index as i32, i as i32];
            }
            map.insert(target - nums[i], i);
        }
        return vec![];
    }
}

fn main() {
    println!("{:?}", Solution::two_sum(vec![2, 7, 11, 15], 9));
    println!("{:?}", Solution::two_sum(vec![3, 2, 4], 6));
    println!("{:?}", Solution::two_sum(vec![3, 3], 6));
}

File: codingpuzzles/1_two_sum.cc
#include <map>
#include <iostream>
#include <vector>

using namespace std;

class Solution
{
public:
    vector<int> twoSum(vector<int> &nums, int target)
    {
        vector<int> res;
        map<int, int> m;

        for (int i = 0; i < nums.size(); i++)
        {
            if (m.find(nums[i]) != m.end())
            {
                res = {m[nums[i]], i};
                return res;
            }
            m[target - nums[i]] = i;
        }

        return res;
    }
};

int main()
{
    // Solution solution();
    vector<int> vec = {2, 7, 11, 15};
    for_each(vec.begin(), vec.end(), [](int &i)
             { printf("%d\n", i); });
    vector<int> res = Solution().twoSum(vec, 9);
    for_each(res.begin(), res.end(), [](int &i)
             { printf("%d\n", i); });
    return 0;
}